v2/internal/reconcilers/reconcile_policy.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package reconcilers
+
+import "github.com/pkg/errors"
+
+// ReconcilePolicyAnnotation describes the reconcile policy for the resource in question.
+// A reconcile policy describes what action (if any) the operator is allowed to take when
+// reconciling the resource.
+// If no reconcile policy is specified, the default is "run"
+const ReconcilePolicyAnnotation = "serviceoperator.azure.com/reconcile-policy"
+
+type ReconcilePolicy string
+
+const (
+	// ReconcilePolicyRun instructs the operator to run all normal reconciliation actions.
+	// This is the default policy when no policy is specified.
+	ReconcilePolicyRun = ReconcilePolicy("run")
+
+	// ReconcilePolicySkip instructs the operator to skip all reconciliation actions. This includes creating
+	// the resource.
+	ReconcilePolicySkip = ReconcilePolicy("skip")
+
+	// ReconcilePolicySkipDelete instructs the operator to skip deletion of resources in Azure. This allows
+	// deletion of the resource in Kubernetes to go through
+	ReconcilePolicySkipDelete = ReconcilePolicy("skip-delete") // TODO: Google config connector calls this deletePolicy: abandon
+)
+
+// ParseReconcilePolicy parses the provided reconcile policy.
+func ParseReconcilePolicy(policy string) (ReconcilePolicy, error) {
+	switch policy {
+	case "":
+		return ReconcilePolicyRun, nil
+	case string(ReconcilePolicyRun):
+		return ReconcilePolicyRun, nil
+	case string(ReconcilePolicySkip):
+		return ReconcilePolicySkip, nil
+	case string(ReconcilePolicySkipDelete):
+		return ReconcilePolicySkipDelete, nil
+	default:
+		// Defaulting to skip. The user is attempting to configure policy but has done it wrong,
+		// if we default to Run we may inadvertently modify their object
+		return ReconcilePolicySkip, errors.Errorf("%q is not a known reconcile policy", policy)
```

Comment:
However, this also makes so that as soon as the `serviceoperator.azure.com/reconcile-policy` annotation is present we skip, no matter the string value. That also makes the verb `skip` unnecessary (as if it doesn't matter if I put `skip` or `sk1p` or `piks` in there). I think this has great implications that as soon as the annotation is present, the operator skips reconcile.

That is fine, I would just be more explicit in the docs about that. Essentially there's three states:
1. I don't set anything and the operator reconciles
2. I set the `serviceoperator.azure.com/reconcile-policy` annotation with any value and the operator doesn't reconcile.
3. I set `serviceoperator.azure.com/reconcile-policy: skip-delete` and the operator reconciles but skips delete.

Given the name of the annotation however, I probably wouldn't expect that the operator skips reconciling as soon as the annotation is set. I'd probably expect that if the annotation was called `serviceoperator.azure.com/skip-reconcile`. But since it's called `serviceoperator.azure.com/reconcile-policy`, I'd probably expect that I have to set the correct policy to achieve the behavior I want (either `skip` or `skip-delete`).

Reply:
Agree that an empty `reconcile-policy` doing `skip` probably doesn't make sense. I'm going to change this behavior and default to `manage` (aka behave normally). That seems to be the standard behavior if you get an annotation like this "wrong".

If we really want to enforce that people don't set a "wrong" value here we could prevent it at the webhook level (which has the advantage of being able to actually tell them they did it wrong as opposed to just doing something they didn't expect), but that seems overkill for now. We could investigate doing that in the future if there's actual desire for it from users.